cola/widgets/startup.py

```diff
@@ -158,24 +167,58 @@ def __init__(self, context, parent=None):
         if list_mode == 'list':
             self.tab_bar.setCurrentIndex(1)
 
+    def _setup_model(self, model, items, repos):
+        item = QtGui.QStandardItem(N_('Open...'))
+        item.setEditable(False)
+        item.setIcon(icons.open_directory())
+        model.appendRow(item)
+
+        normalize = display.normalize_path
+
+        added = set()
+        builder = BuildItem(self.context)
+        for repo in repos:
+            path = normalize(repo['path'])
+            name = normalize(repo['name'])
+            if path in added:
+                continue
+            added.add(path)
+
+            item = builder.get(path, name)
+            model.appendRow(item)
+            items.append(item)
+
     def tab_changed(self, idx):
-        bookmarks = self.bookmarks
-        if idx == 0:
-            bookmarks.setViewMode(QtWidgets.QListView.IconMode)
-            bookmarks.setIconSize(make_size(defs.medium_icon))
-            bookmarks.setGridSize(make_size(defs.large_icon))
-            list_mode = 'folder'
-            for item in self.items:
-                path = item.data(Qt.UserRole)
-                item.setText(self.short_paths.get(path, path))
-        else:
-            bookmarks.setViewMode(QtWidgets.QListView.ListMode)
-            bookmarks.setIconSize(make_size(defs.default_icon))
-            bookmarks.setGridSize(QtCore.QSize())
-            list_mode = 'list'
-            for item in self.items:
-                path = item.data(Qt.UserRole)
-                item.setText(path)
+        models_items = [
+            (self.bookmarks, self.bookmark_items),
+            (self.recent, self.recent_items),
+        ]
+        for model, items in models_items:
+            if idx == 0:
+                mode = QtWidgets.QListView.IconMode
+                icon_size = make_size(defs.medium_icon)
+                grid_size = make_size(defs.large_icon)
+                list_mode = 'folder'
+                name_mode = ICON_MODE
+                rename_enabled = True
+            else:
+                mode = QtWidgets.QListView.ListMode
+                icon_size = make_size(defs.default_icon)
+                grid_size = QtCore.QSize()
+                list_mode = 'list'
+                name_mode = LIST_MODE
+                rename_enabled = False
+
+            self.bookmarks.change_rename_enabled(rename_enabled)
+            self.recent.change_rename_enabled(rename_enabled)
+
+            model.setViewMode(mode)
+            model.setIconSize(icon_size)
+            model.setGridSize(grid_size)
+            for item in items:
+                if isinstance(item, PromptWidgetItem):
+                    item_name = item.change_path_text(name_mode)
+                    model.rename_entry(item, item_name)
```

Comment:
I'm confused about this part -- is this going and calling `rename_entry()` on each item?

In theory that's renaming all of the repos. Am I reading that correctly?

I recall that switching between modes was pretty slow.. if so, this might explain why. It's actually renaming things to their same values and saving the settings file on each line in this loop.

That seems.. not good. That makes me think we shouldn't even allow renaming in this view. What do you think?

Reply:
Yes, it is corrent. And I think so too.

When change Tabs between "Folder" and "List", item text should be changed(`name` <=> `path`).
I added `rename_entry()` here because somehow `item.setText()` in `change_path_text()` emits `itemChnaged` signal. So, `context.settings.bookmarks` and `context.settings.recent` get changed.

To keep consistency, call `rename_entry()` after `change_path_text()` call.
But, by disabling renaming, `itemChange` signal becomes unnecessarry and this part gets to be able to eliminated.